Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,194 @@
+                        and ufo_component_id in glyph.lib[OBJECT_LIBS_KEY]
+                        and (
+                            TRUETYPE_ROUND_KEY
+                            in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                            or TRUETYPE_METRICS_KEY
+                            in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                        )
+                    ):
+                        component_lib = glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+
+                        c.flags &= ~ROUND_XY_TO_GRID
+                        if component_lib.get(TRUETYPE_ROUND_KEY, False):
+                            c.flags |= ROUND_XY_TO_GRID
+
+                        c.flags &= ~USE_MY_METRICS
+                        if component_lib.get(TRUETYPE_METRICS_KEY, False):
+                            if use_my_metrics_comp:
+                                logger.warning(
+                                    "Ignoring USE_MY_METRICS flag on component "
+                                    f"'{ufo_component_id}' because it has been set on "
+                                    f"component '{use_my_metrics_comp}' already."
+                                )
+                            else:
+                                c.flags |= USE_MY_METRICS
+                                use_my_metrics_comp = ufo_component_id
+
+                    # We might automatically set the flags if no data is present,
+                    # but:
+                    # - https://github.com/googlefonts/ufo2ft/pull/425 recommends
+                    #   against setting the ROUND_XY_TO_GRID flag
+                    # - USE_MY_METRICS has been set already by
+                    #   outlineCompiler.OutlineTTFCompiler.autoUseMyMetrics
+
+                    if i == 0 and TRUETYPE_OVERLAP_KEY in glyph.lib:
+                        # Set OVERLAP_COMPOUND on the first component only
+                        c.flags &= ~OVERLAP_COMPOUND
+                        if glyph.lib.get(TRUETYPE_OVERLAP_KEY, False):
+                            c.flags |= OVERLAP_COMPOUND
+
+    def compile_maxp(self):
+        maxp = self.font["maxp"]
+        ttdata = self.ufo.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata:
+            for name in (
+                "maxStorage",
+                "maxFunctionDefs",
+                "maxInstructionDefs",
+                "maxStackElements",
+                # "maxSizeOfInstructions",  # Is recalculated below
+                "maxZones",
+                "maxTwilightPoints",
+            ):
+                value = ttdata.get(name, None)
+                if value is not None:
+                    setattr(maxp, name, value)
+
+        # Recalculate maxp.maxSizeOfInstructions
+        sizes = [
+            len(glyph.program.getBytecode())
+            for glyph in self.font["glyf"].glyphs.values()
+            if hasattr(glyph, "program")
+        ] + [0]
```

Comment:
instead of `+ [0]` use `max(sizes, default=0)`

Reply:
Initially, I had all the hinting compilation stuff in the postProcessor, but a recent optimization (#486) caused the postProcessor to run only on a finished VF, which caused my code to not apply cvt variations. I moved the CVT compilation to the outlineCompiler to build a cvt for each input font.

Running on the VF once is fine for the hinting, because everything except cvt must be identical in all masters. And it should save some processing time.

Reply:
I ran some comparisons with compiling in the outlineCompiler vs. postProcessor:

No hinting: 43.074s
Compilation in outlineCompiler: 43.797s
Compilation in postProcessor: 43.597s

(Thats the complete processing time to build a VF from 4 master UFOs, 1500 glyphs)

No big difference there. It may be a bigger difference when there is a large number of masters, that’s why I tend to keep the hinting compilation in postProcessor.

The additional font compilation/reload time for the "hack" is about 380 ms for a 1500 glyph variable font. The raw hinting compile time is about 280 ms per master on my machine. So for more than 2 masters, the compilation in postProcessor should reduce the total processing time if those numbers scale.